clients/unshaded/src/main/java/tachyon/client/block/BlockStoreContext.java

```diff
@@ -34,15 +36,23 @@
 /**
  * A shared context in each client JVM for common block master client functionality such as a pool
  * of master clients and a pool of local worker clients. Any remote clients will be created and
- * destroyed on a per use basis. This class is thread safe.
+ * destroyed on a per use basis.
+ *
+ * NOTE: The context maintains a pool of block master clients and a pool of block worker clients
+ * that are already thread-safe. Synchronizing {@link BlockStoreContext} methods could lead to
+ * deadlock: thread A attempts to acquire a client when there are no clients left in the pool and
+ * blocks holding a lock on the {@link BlockStoreContext}, when thread B attempts to release a
+ * client it owns, it is unable to do so, because thread A holds the lock on
+ * {@link BlockStoreContext}.
```

Comment:
question: I see it is necessary to avoid deadlock by keeping `releaseMasterClient` out of `synchronized` so a client can always return acquired master client. However, it is really correct by removing all `synchronized` keywords here? Do we require some kind of atomicity for any of those methods?